router/batchrouter/asyncdestinationmanager/marketo-bulk-upload/apiService.go

```diff
@@ -0,0 +1,222 @@
+	PollImportStatus(importId string) (*MarketoResponse, *APIError)
+	GetLeadStatus(url string) ([]map[string]string, *APIError)
+}
+
+type MarketoAPIService struct {
+	logger       logger.Logger
+	statsFactory stats.Stats
+	httpClient   *http.Client
+	munchkinId   string
+	authService  MarketoAuthServiceInterface
+}
+
+type APIError struct {
+	StatusCode int64
+	Category   string
+	Message    string
+}
+
+func (m *MarketoAPIService) checkForCSVLikeResponse(resp *http.Response) bool {
+	// check for csv like response by checking the headers
+	respHeaders := resp.Header
+	return respHeaders.Get("Content-Type") == "text/csv;charset=UTF-8"
+}
+
+func (m *MarketoAPIService) attemptImport(uploadURL, csvFilePath, deduplicationField string, uploadTimeStat stats.Measurement) (string, *APIError) {
+	token, err := m.authService.GetAccessToken()
+	if err != nil {
+		return "", &APIError{StatusCode: 500, Category: "Retryable", Message: "Error in fetching access token"}
+	}
+
+	startTime := time.Now()
+	resp, err := sendHTTPRequest(uploadURL, csvFilePath, token, deduplicationField)
+	uploadTimeStat.Since(startTime)
+
+	if err != nil {
+		return "", &APIError{StatusCode: 500, Category: "Retryable", Message: "Error in sending request"}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", &APIError{StatusCode: int64(resp.StatusCode), Category: "Retryable", Message: "Error in sending request"}
+	}
+
+	responseBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", &APIError{StatusCode: 500, Category: "Retryable", Message: "Error in reading response body"}
+	}
+
+	var marketoResponse MarketoResponse
+	err = json.Unmarshal(responseBody, &marketoResponse)
+	if err != nil {
+		return "", &APIError{StatusCode: 500, Category: "Retryable", Message: "Error in parsing response body"}
+	}
+
+	statusCode, category, errorMessage := parseMarketoResponse(marketoResponse)
+	if category == "Success" {
+		return marketoResponse.Result[0].ImportID, nil
+	}
+
+	return "", &APIError{StatusCode: statusCode, Category: category, Message: errorMessage}
+}
+
+func (m *MarketoAPIService) ImportLeads(csvFilePath, deduplicationField string) (string, *APIError) {
+	uploadTimeStat := m.statsFactory.NewTaggedStat("async_upload_time", stats.TimerType, map[string]string{
+		"module":   "batch_router",
+		"destType": "MARKETO_BULK_UPLOAD",
+	})
+
+	uploadURL := fmt.Sprintf("https://%s.mktorest.com/bulk/v1/leads.json", m.munchkinId)
+
+	// Initial attempt
+	importID, apiError := m.attemptImport(uploadURL, csvFilePath, deduplicationField, uploadTimeStat)
+	if apiError == nil {
+		return importID, nil
+	}
+
+	// If we get a token refresh error, retry once
+	if apiError.Category == "RefreshToken" {
+		m.logger.Info("Token refresh required. Retrying import after fetching new token.")
+		time.Sleep(5 * time.Second) // Wait for 5 seconds before retrying
+		return m.attemptImport(uploadURL, csvFilePath, deduplicationField, uploadTimeStat)
+	}
+
+	return "", apiError
```

Comment:
Can we add a stat here to monitor how many times the retryable error occurred so that we can take steps if any unnecessary retries happen?

Reply:
But wont we have this in jobs db - and I believe router should have such metrics based on jobs db